### PR TITLE
feat(model): add Ling 2.0 / BailingMoeV2 (mini, flash, 1T) (#2242)

### LINKED ISSUE #2242: Support inclusionAI/Ling-mini/flash/1T-2.0
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Hayden727**
# What does this PR do?

Adds support for the `inclusionAI/Ling-2.0` MoE family (`BailingMoeV2ForCausalLM`):
**Ling-mini-2.0** (16B-A1.4B), **Ling-flash-2.0** (100B-A6B), and **Ling-1T** (1T-A50B).
All three variants share the same architecture and are handled by a single implementation.

Closes #2242.

# Changelog

- Add `nemo_automodel/components/models/ling_v2/` with `BailingMoeV2Config`, `BailingMoeV2ForCausalLM`, attention layer (GQA + per-head QK-RMSNorm + half RoPE), and state-dict adapter (HF↔native).
- Register `BailingMoeV2ForCausalLM` arch and `bailing_moe` custom config in `nemo_automodel/_transformers/registry.py`.
- Reuse the framework's existing sigmoid + grouped-topk + per-expert-bias `Gate` (DeepSeek-V3-style) — no new MoE routing code.
- Reuse `gpt_oss/rope_utils.py`'s `partial_rotary_factor` support — no new RoPE code.
- Derive `partial_rotary_factor` from `rotary_dim` when present (Ling-1T uses `rotary_dim=64`; mini/flash use `partial_rotary_factor=0.5`).
- Add LoRA SFT example recipe: `examples/llm_finetune/ling/ling_mini_2_0_squad.yaml`.
- Add model-coverage page: `docs/model-coverage/llm/inclusionai/ling-2.md`.
- Add 16 unit tests including HF↔native state-dict round-trip and per-variant GPU smoke tests for mini / flash / 1T shapes.
- Add manual parity script (`tests/unit_tests/models/ling_v2/parity_ling_v2.py`) implementing the 3 levels from the parity-testing skill, plus a single-GPU real-checkpoint smoke (`_real_forward_smoke.py`).

# Validation

Run inside `nvcr.io/nvidia/nemo-automodel:26.04` on a single H100 80GB:

- **16 unit tests** pass (CPU + GPU; covers all three variant shapes incl. Ling-1T's `first_k_dense_replace=4` and the `rotary_dim`→`partial_rotary_factor` derivation).
- **Level 1** state-dict round-trip on the real 16B `inclusionAI/Ling-mini-2.0` checkpoint: 14813 HF tensors → 299 grouped native → 14813 HF, **zero diff**.
- **Level 2** half-RoPE component parity vs a GPT-NeoX-style reference: `max_diff=0`, `cos=1.0`.
- **Real-checkpoint forward** on H100: 16.26B params loaded, `load_state_dict` reports `missing=0 unexpected=0`, GPU memory 32.5 GB (matches bf16), logits finite, top-1 argmax shows 22 distinct tokens over 32 positions (non-degenerate).
- **Adjacent `tests/unit_tests/_transformers/`** suite (366 tests): no regression.
- **Recipe YAML** accepted by `tools/lint_example_yamls.py` (the CI validator that runs in `validate-recipe-configs.yml`).
- **Level 3** end-to-end HF logits parity is **deferred**: HF's bundled `modeling_bailing_moe_v2.py` imports `is_torch_fx_available` which was removed in `transformers>=5`. To be revisited once upstream Ling patches that or via a `transformers<5` shim.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

# Additional Information

- Related to #2242
- External contributor: please post `/ok to test <full-sha>` so `copy-pr-bot` can run CI.

### CODE REVIEW

**Comment by HuiyingLi** [comment]
Thank you @Hayden727 !
Could you please share the training log/curve for the yaml examples/llm_finetune/ling/ling_mini_2_0_squad.yaml ?
Appreciate it!

**Comment by Hayden727** [comment]
Hi @HuiyingLi — here is the training curve for `examples/llm_finetune/ling/ling_mini_2_0_squad.yaml`.

## Setup

- Container: `nvcr.io/nvidia/nemo-automodel:26.04`
- Hardware: 2 × H100 80GB (`CUDA_VISIBLE_DEVICES=0,1`, `NCCL_NVLS_ENABLE=0`)
- Run command: `automodel examples/llm_finetune/ling/ling_mini_2_0_squad.yaml --nproc-per-node 2 --distributed.ep_size=2 --checkpoint.enabled=false`
- All 500 steps + 5 validation passes finished in **~9 minutes** of wall-clock training.

## Curve

Ling-mini-2.0 LoRA SFT on SQuAD — 500 steps

## Numbers

| Metric | Value |
|---|---|
| Initial train loss | 12.21 |
| Final train loss (step 500) | **0.1004** |
| Final val loss (step 500) | **0.1091** |
| LR schedule | linear warmup 50 steps → cosine decay |
| Peak LR | 1.0e-04 |
| Min LR | 1.0e-06 |
| Per-rank peak GPU memory | ~21 GiB |
| Throughput (global) | 5,000 – 8,000 tokens/sec |

Convergence is clean: loss drops from ~12 to <0.5 by step 40, plateaus around 0.1, and val loss tracks train loss with no overfitting through 500 steps.

Full raw stdout log (213 KB, has every step's metrics) is at:
https://github.com/Hayden727/Automodel/blob/ling-v2-assets/mini_500.log

## YAML changes I made before this run

While preparing this run I caught two issues in the example YAML — both pushed to the PR branch in commit [`a0e7913…`](https://github.com/NVIDIA-NeMo/Automodel/pull/2255/commits) (will follow up with a separate commit):

1. `validation_dataset.num_samples_limit: 128` → renamed to `limit_dataset_samples: 128` to match `make_squad_dataset`'s signature.
2. `distributed.ep_size: 1` → `4` (the default since the framework's HF state-dict adapter for grouped MoE experts requires an `ep` mesh dimension; users with a different GPU count override via `--distributed.ep_size=N`).
3. Switched recipe default to `attn: sdpa` (works without TE installed) and added a `lr_scheduler` block with the standard cosine + warmup schedule.

## Note on `Ling-flash-2.0` (100 B) and `Ling-1T`

Same architecture, same code path — both arch-validated via per-variant tiny-config GPU smoke tests already in this PR.  I was not able to run real-checkpoint training on this 8×H100 80GB box for either:

- **Ling-flash-2.0**: needs ~200 GB checkpoint download + 4-8 GPUs free simultaneously; the host's other tenants kept grabbing GPUs mid-run.  Feasible on a dedicated 8-GPU node.
- **Ling-1T**: total bf16 weights ≈ 2 TB > 640 GB total GPU memory on this box — the model physically cannot be held even sharded.  A real run would need ≥32 H100s.

Happy to run them on dedicated hardware if NVIDIA has a slot, or wait for follow-up PRs that focus on the larger variants.

**Comment by Hayden727** [comment]
@HuiyingLi — yes the previous curve was on SQuAD (answer-only masked CE, so the labels are mostly rare named entities; that's where the ~12 initial loss came from on a base instruct model that's never seen the SQuAD format).  Re-ran the same recipe on HellaSwag and the initial loss lands right where you predicted.

## Setup

- Same 2 × H100 80GB box (`CUDA_VISIBLE_DEVICES=0,1`, `NCCL_NVLS_ENABLE=0`)
- New recipe: [`examples/llm_finetune/ling/ling_mini_2_0_hellaswag.yaml`](https://github.com/NVIDIA-NeMo/Automodel/pull/2255/files) (added in commit `1889664…` on this PR — will push it to the PR branch alongside this comment)
- 500 steps + 5 validation passes in **~7.5 min** wall-clock

## Curve

Ling-mini-2.0 LoRA SFT on HellaSwag — 500 steps

## Numbers

| Metric | SQuAD (previous run) | HellaSwag (new) |
|---|---|---|
| Initial train loss (step 0) | 12.21 | **3.35** |
| Steps 1–4 train loss | 12.34 / 9.97 / 9.04 / 8.98 | **3.13 / 3.23 / 3.11 / 3.45** |
| Final train loss (step 500) | 0.10 | 2.08 |
| Final val loss (step 500) | 0.11 | 2.21 |
| Wall-clock | ~9 min | ~7.5 min |

HellaSwag's initial 3.35 matches your "2–3 ish" expectation — the gap between the two datasets is exactly the format-mismatch story (SQuAD's CE is over short answers with rare tokens; HellaSwag's CE is over fluent natural-language continuations the base model has plenty of mass on).

Raw log: https://github.com/Hayden727/Automodel/blob/ling-v2-assets/hellaswag_500.log

I'll push the new YAML to the PR branch in a follow-up commit.

**Comment by Hayden727** [comment]
@HuiyingLi — pushed two more example recipes for the larger Ling-2.0 variants so you can validate them on the right hardware.  Both committed in `f0a17556…` on this PR branch.

| Recipe | Min hardware | Notes |
|---|---|---|
| [`ling_flash_2_0_lora.yaml`](https://github.com/NVIDIA-NeMo/Automodel/pull/2255/files#diff-flash) | 8 × H100 80GB (1 node) | FSDP2 + EP=8 + LoRA + activation_ckpt; TE+DeepEP MoE backend.  Modelled after `qwen3_moe_30b_lora.yaml`. |
| [`ling_1t_lora_pp.yaml`](https://github.com/NVIDIA-NeMo/Automodel/pull/2255/files#diff-1t) | 32 × H100 80GB (4 nodes) | FSDP2 + **PP=4** + EP=8 + LoRA + activation_ckpt; interleaved 1f1b, `layers_per_stage=5` (80 layers / pp 4 / 4 virtual stages).  Modelled after `deepseek_v32_hellaswag_pp.yaml`. |

Both load the model via `BailingMoeV2Config.from_pretrained`, so the per-variant fields (`first_k_dense_replace=1` for flash vs `=4` for 1T, `rotary_dim=64` on 1T, etc.) are picked up from each checkpoint's `config.json` — no per-recipe overrides needed.

Validation status (transparent):
- ✅ Both YAMLs pass `tools/lint_example_yamls.py`.
- ✅ The size-specific code paths (1T's 4 dense layers + `rotary_dim`) are exercised by the per-variant tiny-shape GPU smoke tests already in this PR (`tests/unit_tests/models/ling_v2/test_ling_v2_variants.py`).
- ❌ I did not run either end-to-end — the shared box I have only consistently exposes 2-4 H100s at a time, not the 8 / 32 these need.  Per our offline chat you'll handle the real-checkpoint runs.

Let me know if the parallelism layout / batch sizes need tuning for your slot — happy to push fixes.

**Comment by HuiyingLi** [comment]
Hi @Hayden727 , could we also add SFT recipes. They might carry different parallelism setups. Thank you~

**Comment by Hayden727** [comment]
@HuiyingLi — pushed three full-SFT recipes in commit `2ab2ae63…`:

| Recipe | Min HW | Parallelism | Notes |
|---|---|---|---|
| [`ling_mini_2_0_sft.yaml`](https://github.com/NVIDIA-NeMo/Automodel/pull/2255/files) | **8 × H100 80GB** (1 node) | FSDP2 + EP=8 | Same topology as `qwen3_moe_30b_hellaswag.yaml`; 16B Adam state in fp32 is ~190 GB ⇒ ~24 GB/rank. |
| [`ling_flash_2_0_sft.yaml`](https://github.com/NVIDIA-NeMo/Automodel/pull/2255/files) | **32 × H100 80GB** (4 nodes) | FSDP2 + EP=32 | 100B Adam state in fp32 is ~1.2 TB; 32 ranks give ~37 GB/rank. Includes `moe.reshard_after_forward` + `wrap_outer_model` for memory headroom. |
| [`ling_1t_sft.yaml`](https://github.com/NVIDIA-NeMo/Automodel/pull/2255/files) | **256 × H100 80GB** (32 nodes) | FSDP2 + PP=4 + EP=64 | Modelled after `deepseek_v32_hellaswag_pp.yaml` (671B / same 32-node H100 topology); `interleaved1f1b` with `layers_per_stage=5` (80 layers / pp=4 / 4 virtual). |

All three pass `tools/lint_example_yamls.py` and the model-coverage table now lists both LoRA + SFT variants with explicit "Min HW" columns.

Key SFT-vs-LoRA changes per recipe:
- `peft:` block removed (every param trainable)
- `optimizer.lr` lowered (2e-5 / 1e-5 / 5e-6 vs LoRA's 1e-4)
- `global_batch_size` raised for stability (256 / 1024 / 512)
- `activation_checkpointing: true` mandatory at all scales
- 1T inherits the DSv3.2 `attn: sdpa, linear: te` combo per the same-scale reference recipe

Same hardware-reality caveats apply: I can validate end-to-end on `ling_mini_2_0_sft.yaml` (8 H100), but flash and 1T SFT need the multi-node hardware you have access to.  Happy to tune batch sizes / LR / EP layout if your runs hit anything off.

**Comment by HuiyingLi** [comment]
@Hayden727 flash runs
image
image

**Comment by Hayden727** [comment]
@HuiyingLi — thanks for catching both!  Fixed in commit `9617bbdb…`.

## Root cause of the `apply_rotary_emb` crash

The traceback was at `torch.cat((o1, o2, x_pass), dim=-1)` inside `apply_rotary_emb`, with shapes broadcasting in a wrong direction:

```
File ".../distributed/pipelining/hf_utils.py", line 153, in pipeline_forward
    position_embeddings = rotary_emb(hidden_states, position_ids)   # <-- assumes HF style
File ".../models/gpt_oss/rope_utils.py", line 131, in forward
    query = apply_rotary_emb(query, cos, sin)                       # <-- but our rotary rotates q/k directly
```

`patch_hf_model_for_pp` replaces our model's forward with a generic HF-style `pipeline_forward` that calls `self.model.rotary_emb(hidden_states, position_ids)` expecting a `(cos, sin)` tuple back.  Our `BailingMoeV2` inherits the gpt_oss convention where `RotaryEmbedding.forward(query, key)` actually rotates q/k tensors — so `hidden_states (B, S, H)` and `position_ids (B, S)` get fed into the rotation math and the shapes diverge by the time we reach the final `cat`.  That's exactly the `Expected size 168 but got size 1` you saw.

## Fix

1. **Code** — added `_pp_keep_self_forward: bool = True` on `BailingMoeV2ForCausalLM` (matches Qwen3_5MoE / KimiVL / Kimi-K2.5-VL).  The PP splitter checks this and skips `patch_hf_model_for_pp` entirely, preserving our freqs_cis-based forward.
2. **YAMLs** — both 1T recipes now also explicitly carry `patch_inner_model: false` and `patch_causal_lm_model: false` as belt + suspenders so a recipe override can't reintroduce the bug.
3. **Test** — new `test_pp_keep_self_forward_is_declared` in `tests/unit_tests/models/ling_v2/test_ling_v2_variants.py` guards the class flag.  17 unit tests pass total (was 16).

## 1T LoRA topology

Updated `ling_1t_lora_pp.yaml` to your validated topology:

| | Old | New |
|---|---|---|
| GPUs | 32 (4 nodes × 8) | **64 (8 nodes × 8)** |
| PP | 4 | **8** |
| EP | 8 | 8 |
| `layers_per_stage` | 5 | **2** (80 layers / pp=8 / virtual=5) |

Docs table updated accordingly.

1T SFT (`ling_1t_sft.yaml`) still uses the 256-GPU / 32-node topology copied from `deepseek_v32_hellaswag_pp.yaml` since the Adam state footprint dominates there — happy to spread that further if your run says otherwise.

Need NVIDIA to post `/ok to test 9617bbdb` to re-trigger CI.

**Comment by HuiyingLi** [comment]
@Hayden727 Thank you for the fix! Could you please also resolve the conflicts

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review of .md files. No copyedits needed. LGTM.

**Comment by HuiyingLi** [comment]
## We observe DeepEP timeout on Ling-1T SFT and LoRA runs

@Hayden727 Root cause for the Ling-1T DeepEP failure:

Ling-1T has `hidden_size=8192`, which crosses a DeepEP intranode TMA shared-memory limit. The failing DeepEP path checks:

`half_hidden_bytes + sizeof(uint64_t) <= kNumTMABytesPerWarp`

For bf16 hidden size 8192, `half_hidden_bytes` is already 8192 bytes, so the extra 8-byte metadata pushes it to 8200 bytes. That exceeds the 8192-byte per-warp TMA budget. This is why smaller MoE models with hidden sizes like 4096 or 7168 work, but Ling-1T does not with `dispatcher: deepep`.

The practical workaround is to use `dispatcher: hybridep`.

One additional tuning was needed for the full SFT topology. With `ep_size=64`, `pp_size=4`, and `NUM_OF_HYBRID_EP_RANKS_PER_NVLINK_DOMAIN=8`, each EP group spans 8 nodes. HybridEP’s default multi-node backward combine config requests slightly more dynamic shared memory than H100 allows:

- H100 opt-in shared memory limit: 232448 bytes
- default HybridEP backward combine request: 233984 bytes

Setting:

`NUM_OF_STAGES_G2S_COMBINE_API=4`

reduces the HybridEP backward combine shared-memory request to about 200832 bytes, which fits under the H100 limit.

So the recommended Ling-1T config is:
- use `dispatcher: hybridep`
- keep `rope_fusion: false`
- set `NUM_OF_HYBRID_EP_RANKS_PER_NVLINK_DOMAIN=8`
- for multi-node EP64 SFT, also set `NUM_OF_STAGES_G2S_COMBINE_API=4`

**Comment by HuiyingLi** [comment]
@Hayden727  
1T 32nodes SFT
image

1T ep8pp8 8nodes
image

**Review by HuiyingLi — APPROVED** [review]
LGTM, thank you so much!

File: examples/llm_finetune/ling/ling_1t_lora_pp.yaml [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  weight_decay: 0.01
+  betas: [0.9, 0.95]
+  eps: 1e-8
+
+lr_scheduler:
+  lr_decay_style: cosine
+  min_lr: 1.0e-6
+
+ci:
+  recipe_owner: Hayden727
```
Comment (akoumpa) [inline examples/llm_finetune/ling/ling_1t_lora_pp.yaml:128]: @HuiyingLi should we update this?
